core/transaction-pool/src/lib.rs

```diff
@@ -20,10 +20,246 @@
+use futures::Future;
+
+use client::{
+	Client,
+	light::fetcher::Fetcher,
+	runtime_api::TaggedTransactionQueue,
+};
+use primitives::{Blake2Hasher, H256};
+use sr_primitives::{
+	serde::Serialize,
+	generic::BlockId,
+	traits::{Block as BlockT, Member, ProvideRuntimeApi},
+};
+
+use txpool::{BestIterator, EventStream, Options, watcher::Watcher};
+
+/// Extrinsic hash type for a pool.
+pub type ExHash<P> = <P as TransactionPool>::Hash;
+/// Block hash type for a pool.
+pub type BlockHash<P> = <<P as TransactionPool>::Block as BlockT>::Hash;
+/// Extrinsic type for a pool.
+pub type ExtrinsicFor<P> = <<P as TransactionPool>::Block as BlockT>::Extrinsic;
+
+/// Transaction pool interface.
+pub trait TransactionPool: Send + Sync {
+	/// Block type.
+	type Block: BlockT;
+	/// Transaction Hash type.
+	type Hash: Hash + Eq + Member + Serialize;
+	/// Error type.
+	type Error: From<txpool::error::Error> + txpool::error::IntoPoolError;
+
+	/// Returns a future that imports a bunch of unverified extrinsics to the pool.
+	fn submit_at(
+		&self,
+		at: &BlockId<Self::Block>,
+		xts: impl IntoIterator<Item=ExtrinsicFor<Self>> + 'static,
+		force: bool,
+	) -> Box<dyn Future<Output=Result<Vec<Result<Self::Hash, Self::Error>>, Self::Error>> + Send + Unpin>;
+
+	/// Returns a future that imports one unverified extrinsic to the pool.
+	fn submit_one(
+		&self,
+		at: &BlockId<Self::Block>,
+		xt: ExtrinsicFor<Self>,
+	) -> Box<dyn Future<Output=Result<Self::Hash, Self::Error>> + Send + Unpin>;
+
+	/// Returns a future that import a single extrinsic and starts to watch their progress in the pool.
+	fn submit_and_watch(
+		&self,
+		at: &BlockId<Self::Block>,
+		xt: ExtrinsicFor<Self>,
+	) -> Box<dyn Future<Output=Result<Watcher<Self::Hash, BlockHash<Self>>, Self::Error>> + Send + Unpin>;
+
+	/// Remove extrinsics identified by given hashes (and dependent extrinsics) from the pool.
+	fn remove_invalid(
+		&self,
+		hashes: &[Self::Hash],
+	) -> Vec<Arc<txpool::base_pool::Transaction<Self::Hash, ExtrinsicFor<Self>>>>;
+
+	/// Returns pool status.
+	fn status(&self) -> txpool::base_pool::Status;
+
+	/// Get an iterator for ready transactions ordered by priority
+	fn ready(&self) -> BestIterator<Self::Hash, ExtrinsicFor<Self>>;
+
+	/// Return an event stream of transactions imported to the pool.
+	fn import_notification_stream(&self) -> EventStream;
+
+	/// Returns transaction hash
+	fn hash_of(&self, xt: &ExtrinsicFor<Self>) -> Self::Hash;
+
+	/// Notify the pool about transactions broadcast.
+	fn on_broadcasted(&self, propagations: HashMap<Self::Hash, Vec<String>>);
+
+	/// Returns a future that performs maintenance procedures on the pool.
+	fn maintain(
+		&self,
+		id: &BlockId<Self::Block>,
+		retracted: &[BlockHash<Self>],
+	) -> Box<dyn Future<Output=()> + Send + Unpin>;
+}
+
+/// Basic implementation of transaction pool that can be customized by providing
+/// different PoolApi and Maintainer.
+pub struct BasicTransactionPool<PoolApi, Maintainer, Block>
```

Comment:
Cant' we merge `PollApi` and `Maintainer` somehow? You can't really have `FullChainApi + LightPoolMaintainer`, can you? Alternatively maybe just merge `BasicTransactionPool` with the maintainer implementation? The only drawback would be that you have to duplicate method implementations I guess?

Reply:
Or perhaps let's move `fn maintain` trait to a separate thread, and then have:
```rust
// generic implementation that does not care about Light/Full stuff
impl TransactionPool for BasicTransactionPool<...>  {

}
// Light/Full-specific maintainer implementation
impl TransactionPoolMaintain for BasicTransactionPool<
  LightChainApi<..>
> {
 fn maintain(...) {}
}
```
This will allow us to reduce the amount of code a bit and rid of `Default*PoolMaintainer` implementations.

I think you could also simplify pool creation with by switchin to no-maintainer and using bare functions:
```rust
fn new_full<Client, Block>(client: Arc<Client>) -> BasicTransactionPool<
  FullChainApi<Client, Block>
> where 
  Block: traits::Block,
  Client: ProvideRuntimeApi + HeaderBackend,
{

}
```
You don't really need all of those bounds to create stuff, only when you add behavior (i.e. imlement `TransactionPool` trait).

Reply:
(will leave a top-level comment about that)